File: LAB_02/src/ki032/Smolynskyi/Lab2/Automat.java
/**
 * 
 */
package ki032.Smolynskyi.Lab2;

import java.io.*;

/**
 * Клас {@code Automat} моделює автоматичну зброю.
 * <p>
 * Має три основні складові:
 * <ul>
 *   <li>{@link Barrel} – ствол;</li>
 *   <li>{@link Trigger} – курок;</li>
 *   <li>{@link Magazine} – магазин.</li>
 * </ul>
 * Методи класу ведуть протокол своєї діяльності
 * у текстовий файл <b>automat_log.txt</b>.
 * </p>
 */
public class Automat {
    private Barrel barrel;
    private Trigger trigger;
    private Magazine magazine;
    private PrintWriter logWriter;

    /**
     * Конструктор за замовчуванням.
     * <p>Створює автомат із типовим магазином на 30 набоїв.</p>
     *
     * @throws IOException якщо не вдалося відкрити файл протоколу
     */
    public Automat() throws IOException {
        this(new Barrel(), new Trigger(), new Magazine(30));
    }

    /**
     * Конструктор з параметрами.
     *
     * @param barrel   об’єкт ствола
     * @param trigger  об’єкт курка
     * @param magazine об’єкт магазину
     * @throws IOException якщо не вдалося відкрити файл протоколу
     */
    public Automat(Barrel barrel, Trigger trigger, Magazine magazine) throws IOException {
        this.barrel = barrel;
        this.trigger = trigger;
        this.magazine = magazine;
        logWriter = new PrintWriter(new FileWriter("automat_log.txt", true));
        log("Автомат створено");
    }

    /**
     * Виконує один постріл.
     * Постріл можливий лише за умови наявності патронів у магазині
     * та натиснутого курка.
     */
    public void fireSingle() {
        if (magazine.hasBullets() && trigger.isPressed()) {
            magazine.useBullet();
            log("Постріл! Залишилось куль: " + magazine.getBullets());
        } else {
            log("Неможливо здійснити постріл");
        }
    }

    /**
     * Виконує серію пострілів (чергу).
     *
     * @param count кількість пострілів у черзі
     */
    public void fireBurst(int count) {
        log("Спроба черги на " + count + " пострілів");
        for (int i = 0; i < count; i++) {
            fireSingle();
        }
    }

    /**
     * Перезаряджає магазин.
     *
     * @param bullets кількість патронів для заряджання
     */
    public void reload(int bullets) {
        magazine.reload(bullets);
        log("Магазин перезаряджено. Куль: " + magazine.getBullets());
    }

    /** Натискає на курок. */
    public void pressTrigger() {
        trigger.press();
        log("Курок натиснуто");
    }

    /** Відпускає курок. */
    public void releaseTrigger() {
        trigger.release();
        log("Курок відпущено");
    }

    /**
     * Повертає кількість наявних патронів.
     *
     * @return кількість куль у магазині
     */
    public int getRemainingBullets() {
        return magazine.getBullets();
    }

    /** Повністю спорожнює магазин. */
    public void clearMagazine() {
        magazine.clear();
        log("Магазин спорожнено");
    }

    /**
     * Перевіряє готовність автомата до стрільби.
     * Умови готовності: є кулі та ствол не перегрітий.
     *
     * @return {@code true}, якщо можна стріляти;
     *         {@code false} – якщо ні
     */
    public boolean isReady() {
        boolean ready = magazine.hasBullets() && !barrel.isOverheated();
        log("Перевірка готовності: " + ready);
        return ready;
    }

    /**
     * Коректне завершення роботи з файлом протоколу.
     * <p>Цей метод необхідно викликати перед виходом із програми.</p>
     */
    public void dispose() {
        log("Автомат вимкнено");
        logWriter.close();
    }

    /**
     * Допоміжний метод для запису повідомлення у лог.
     *
     * @param message текст повідомлення
     */
    private void log(String message) {
        logWriter.println(message);
        logWriter.flush();
    }
}

File: ki302_Smolynskyi_Lab9/automat.py
"""
Модуль 'automat'
Містить абстрактний базовий клас 'Automat'.
"""

from abc import ABC, abstractmethod

class Automat(ABC):
    """
    Абстрактний базовий клас 'Automat'.

    Описує базову поведінку автоматичної зброї, таку як
    перезарядка та стрільба. Має абстрактний метод
    для отримання типу зброї.
    """
    
    def __init__(self, magazine_capacity: int):
        """
        Конструктор базового класу.

        Args:
            magazine_capacity (int): Максимальна кількість набоїв.
        """
        self._magazine_capacity = magazine_capacity
        self._bullets_count = 0
        self._log("Створено базовий 'Автомат'")

    def _log(self, message: str):
        """
        Допоміжний "захищений" метод для логування дій.
        
        Args:
            message (str): Повідомлення для виведення.
        """
        print(f"[LOG: {self.__class__.__name__}] {message}")

    @abstractmethod
    def get_weapon_type(self) -> str:
        """
        Абстрактний метод для отримання типу зброї.
        Має бути реалізований у похідних класах.

        Returns:
            str: Назва типу зброї.
        """
        pass

    def reload(self, bullets: int):
        """
        Перезаряджає магазин.

        Args:
            bullets (int): Кількість набоїв для заряджання.
        """
        if bullets > self._magazine_capacity:
            self._bullets_count = self._magazine_capacity
            self._log(f"Перезаряджено. У магазині {self._bullets_count} набоїв (максимум).")
        else:
            self._bullets_count = bullets
            self._log(f"Перезаряджено. У магазині {self._bullets_count} набоїв.")

    def fire_single(self):
        """
        Виконує один постріл.
        """
        if self._bullets_count > 0:
            self._bullets_count -= 1
            self._log(f"Постріл! Залишилось набоїв: {self._bullets_count}")
        else:
            self._log("Неможливо стріляти: магазин порожній.")

    def get_remaining_bullets(self) -> int:
        """
        Повертає кількість наявних набоїв.

        Returns:
            int: Кількість набоїв.
        """
        return self._bullets_count